### PR TITLE
Refactoring: change the signature of push.Func function

### PR DESCRIPTION
**Pull request by duricanikolic**
#### What this PR does
This PR changes the signature of the `push.Func` function from
```
type Func func(ctx context.Context, req *Request) (*mimirpb.WriteResponse(), error)
```
to
```
type Func func(ctx context.Context, req *Request) error
```
The main reason for this change is that all implementations of `push.Func()` actually return:
- `&mimirpb.WriteResponse()` in the case of no error, or 
- `nil` in the case of error.

This logic has been moved to the topmost callers of `push.Func()`, which are mostly `ingester.Push()` and `distributor.Push()`.

This refactoring is just a preliminary change needed for a bigger refactoring and improving of intester's and distributor's write path error handling.

#### Which issue(s) this PR fixes or relates to

Related to https://github.com/grafana/mimir/issues/1900, https://github.com/grafana/mimir/issues/6008

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/push/push.go [hunk 1]
```diff
@@ -115,7 +115,7 @@ func handler(maxRecvMsgSize int,
 			return &req.WriteRequest, cleanup, nil
 		}
 		req := newRequest(supplier)
-		if _, err := push(ctx, req); err != nil {
+		if err := push(ctx, req); err != nil {
```
Comment (pstibrany) [inline pkg/util/push/push.go:118]: what if `err` is nil? :) looks strange we don't handle that *at all*.

**Review by pstibrany — APPROVED** [review]
thanks!

File: pkg/util/push/push.go (hunk 1 above)
Comment (colega) [reply pkg/util/push/push.go:118]: Sounds like it has always been like this? I guess we rely on default golang HTTP handling.

Comment (duricanikolic) [reply pkg/util/push/push.go:118]: I confirm that it was already like that. In this case "no error" is implicitly translated into an HTTP 200.
I have added a test to show that.

Comment (pstibrany) [reply pkg/util/push/push.go:118]: I agree that this PR is not changing. I've just never noticed it before :)

**Pull request merged**